lib/cadet/assessments/assessments.ex

```diff
@@ -121,15 +121,52 @@ defmodule Cadet.Assessments do
     |> Repo.one()
   end
 
-  def assessment_with_questions_and_answers(id, user = %User{}) when is_ecto_id(id) do
+  def assessment_with_questions_and_answers(
+        assessment = %Assessment{password: nil},
+        user = %User{},
+        nil
+      ) do
+    assessment_with_questions_and_answers(assessment, user)
+  end
+
+  def assessment_with_questions_and_answers(
+        assessment = %Assessment{password: nil},
+        user = %User{},
+        _
+      ) do
+    assessment_with_questions_and_answers(assessment, user)
+  end
+
+  def assessment_with_questions_and_answers(
+        assessment = %Assessment{password: password},
+        user = %User{},
+        given_password
+      ) do
+    cond do
+      Timex.after?(Timex.now(), assessment.close_at) ->
+        assessment_with_questions_and_answers(assessment, user)
+
+      given_password == nil ->
+        {:error, {:forbidden, "Missing Password."}}
+
+      password == given_password ->
+        assessment_with_questions_and_answers(assessment, user)
+
+      true ->
+        {:error, {:forbidden, "Invalid Password."}}
+    end
+  end
+
+  def assessment_with_questions_and_answers(id, user = %User{}, password)
```

Comment:
I just realised, this is actually a bad way to check if an assessment is password protected or not. 

We should do something like

```elixir
def assessment_with_questions_and_answers(id, user = %User{}, _) when is_password_protected(id) do
  ...
end

def assessment_with_questions_and_answers(id, user = %User{}, password) do
  ...
end
```

Reply:
`is_password_protected(id)` how would this work tho you will need to make a db call on id to check. long running functions dont make good guard clauses!

im actually using pattern matching to check if password protected
`def assessment_with_questions_and_answers(
        assessment = %Assessment{password: nil},
        user = %User{},
        nil
      ) do`